### PR TITLE
feat: Implement websocket authentication using api-token

### PR DESCRIPTION
**Pull request by jakoblorz**
When #1263 is merged, private software (soon public and OSS) I wrote will be unable to connect to the websocket using only the `apiToken`. Users will be required to enter their `accessToken` which is highly insecure (under the same argument of "possibly insecure in the future" as in #1263).

I propose that access to the websocket should also be granted by authenticating with the `apiToken`. The changes for that are implemented "adjacent" to the http authentication / token resolve functionality.

https://github.com/outline/outline/blob/48c87a1902f1b97b49d8b6d3bd351bbfc7745413/server/middlewares/authentication.js#L45

While this PR is purely motivated by my own needs, I do think that others may find this useful. Especially as it enables developers to react in near-realtime to changes in outline while web hooks are not available (yet).

### CODE REVIEW

**Comment by tommoor** [comment]
I don't feel great about the repetition between this and the authentication middleware, especially for such sensitive logic. I think it'd make sense to pull it into a shared lib method?

**Comment by jakoblorz** [comment]
@tommoor Will do, I am thinking of a reusable method in /server/middlewares/authentication.js.

File: server/middlewares/authentication.js [hunk 1]
```diff
@@ -8,69 +8,79 @@ import addMonths from 'date-fns/add_months';
...
-        const credentials = parts[1];
-
-        if (/^Bearer$/i.test(scheme)) {
-          token = credentials;
-        }
-      } else {
-        throw new AuthenticationError(
-          `Bad Authorization header format. Format is "Authorization: Bearer <token>"`
-        );
+function extractTokenFromCtx(ctx: Context): any {
```
Comment (tommoor) [inline server/middlewares/authentication.js:11]: I'd like to stick to the "getXForY" pattern on these method names

```suggestion
function getTokenForCtx(ctx: Context): ?string {
```

File: server/middlewares/authentication.js [hunk 2]
```diff
@@ -8,69 +8,79 @@ import addMonths from 'date-fns/add_months';
...
+    // $FlowFixMe
+  } else if (ctx.body && ctx.body.token) {
+    return ctx.body.token;
+  } else if (ctx.request.query.token) {
+    return ctx.request.query.token;
+  } else {
+    return ctx.cookies.get('accessToken');
+  }
+
+  return null;
```
Comment (tommoor) [inline server/middlewares/authentication.js:36]: Would be fine as…

```suggestion
  return;
```

File: server/middlewares/authentication.js [hunk 3]
```diff
@@ -8,69 +8,79 @@ import addMonths from 'date-fns/add_months';
...
+}
+
+export async function extractUserFromToken(token: any): Promise<User> {
```
Comment (tommoor) [inline server/middlewares/authentication.js:39]:
```suggestion
export async function getUserForToken(token: string): Promise<User> {
```

**Pull request closed without merging**